gprofiler/metadata/cloud_metadata.py

```diff
@@ -0,0 +1,154 @@
+    if life_cycle_response is None or metadata_response is None:
+        return None
+    instance = metadata_response.json()
+    return AwsInstanceMetadata(
+        provider="aws",
+        region=instance["region"],
+        zone=instance["availabilityZone"],
+        instance_type=instance["instanceType"],
+        life_cycle=life_cycle_response.text,
+        account_id=instance["accountId"],
+        image_id=instance["imageId"],
+        instance_id=instance["instanceId"],
+    )
+
+
+def get_gcp_metadata() -> Optional[GcpInstanceMetadata]:
+    # Documentation: https://cloud.google.com/compute/docs/storing-retrieving-metadata
+    response = send_request(
+        "http://metadata.google.internal/computeMetadata/v1/instance/?recursive=true",
+        headers={"Metadata-Flavor": "Google"},
+    )
+    if response is None:
+        return None
+    instance = response.json()
+    return GcpInstanceMetadata(
+        provider="gcp",
+        zone=instance["zone"],
+        instance_type=instance["machineType"],
+        preemptible=instance["scheduling"]["preemptible"] == "TRUE",
+        preempted=instance["preempted"] == "TRUE",
+        instance_id=str(instance["id"]),
+        image_id=instance["image"],
+        name=instance["name"],
+    )
+
+
+def get_azure_metadata() -> Optional[AzureInstanceMetadata]:
+    # Documentation: https://docs.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux
+    response = send_request(
+        "http://169.254.169.254/metadata/instance/compute/?api-version=2019-08-15", headers={"Metadata": "true"}
+    )
+    if response is None:
+        return None
+    instance = response.json()
+    image_info = None
+    storage_profile = instance.get("storageProfile")
+    if isinstance(storage_profile, dict):
+        image_reference = storage_profile.get("imageReference")
+        if isinstance(image_reference, dict):
+            image_info = {
+                "image_id": image_reference["id"],
+                "image_offer": image_reference["offer"],
+                "image_publisher": image_reference["publisher"],
+                "image_sku": image_reference["sku"],
+                "image_version": image_reference["version"],
+            }
+
+    return AzureInstanceMetadata(
+        provider="azure",
+        instance_type=instance["vmSize"],
+        zone=instance["zone"],
+        region=instance["location"],
+        subscription_id=instance["subscriptionId"],
+        resource_group_name=instance["resourceGroupName"],
+        resource_id=instance["resourceId"],
+        instance_id=instance["vmId"],
+        name=instance["name"],
+        image_info=image_info,
+    )
+
+
+def send_request(url: str, headers: Dict[str, str] = None) -> Optional[Response]:
+    response = requests.get(url, headers=headers or {}, timeout=METADATA_REQUEST_TIMEOUT)
+    if not response.ok:
+        raise BadResponseCode(response.status_code)
+    return response
```

Comment:
When are you returning `None` exactly? At which situation does `requests` return a timeout